File: serializers/gson/src/main/java/at/stefangeyer/challonge/serializer/gson/adapter/MatchWrapperListWrapperAdapter.java
package at.stefangeyer.challonge.serializer.gson.adapter;

import at.stefangeyer.challonge.model.wrapper.MatchWrapper;
import at.stefangeyer.challonge.model.wrapper.MatchWrapperListWrapper;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class MatchWrapperListWrapperAdapter implements JsonDeserializer<MatchWrapperListWrapper> {

    @Override
    public MatchWrapperListWrapper deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonArray jsonArray = json.getAsJsonArray();
        List<MatchWrapper> list = new ArrayList<>();

        for (JsonElement e : jsonArray) {
            MatchWrapper wrapper = context.deserialize(e, MatchWrapper.class);
            list.add(wrapper);
        }

        return new MatchWrapperListWrapper(list);
    }
}

File: serializers/gson/src/main/kotlin/at/stefangeyer/challonge/serializer/gson/adapter/MatchWrapperListWrapperAdapter.kt
package at.stefangeyer.challonge.serializer.gson.adapter

import at.stefangeyer.challonge.model.wrapper.MatchWrapper
import at.stefangeyer.challonge.model.wrapper.MatchWrapperListWrapper
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import java.lang.reflect.Type

class MatchWrapperListWrapperAdapter : JsonDeserializer<MatchWrapperListWrapper> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): MatchWrapperListWrapper {
        val jsonArray = json.asJsonArray
        val list = mutableListOf<MatchWrapper>()

        for (e in jsonArray) {
            val wrapper = context.deserialize<MatchWrapper>(e, MatchWrapper::class.java)
            list.add(wrapper)
        }

        return MatchWrapperListWrapper(list)
    }
}